src/Agent.Worker/Build/BuildJobExtension.cs

```diff
@@ -38,24 +49,93 @@ private Task PrepareAsync()
             IExecutionContext executionContext = PrepareStep.ExecutionContext;
             var directoryManager = HostContext.GetService<IBuildDirectoryManager>();
 
+            // This flag can be false for jobs like cleanup artifacts.
+            // If syncSources = false, we will not set source related build variable, not create build folder, not sync source.
+            bool syncSources = executionContext.Variables.Build_SyncSources ?? true;
+            if (!syncSources)
+            {
+                Trace.Verbose($"{Constants.Variables.Build.SyncSources} = false, we will not set source related build variable, not create build folder and not sync source");
+                executionContext.Output($"{Constants.Variables.Build.SyncSources} = false, skip get source");
+                return;
+            }
+
             // Get the repo endpoint and source provider.
-            ServiceEndpoint endpoint = null;
-            ISourceProvider sourceProvider = null;
-            if (!TryGetPrimaryEndpointAndSourceProvider(executionContext, out endpoint, out sourceProvider))
+            if (!TryGetPrimaryEndpointAndSourceProvider(executionContext, out _sourceEndpoint, out _sourceProvider))
```

Comment:
If these are instance-level fields, and if TryGetPrimaryEndpointAndSourceProvider is only called from this one place, then consider:
1) Not passing _sourceEndpoint and _sourceProvider. Let the method set the instance-level fields directly.
2) Changing the name of the method to something simpler like TrySetPrimaryProviderInfo()